### PR TITLE
Update/shpc version biocontainers repo

### PR DESCRIPTION
**Pull request by marcodelapierre**
@dipietrantonio this is a major SHPC version update, as it provides full support for a registry with 8000+ biocontainers, in a major hike in support for our community of life scientists! -- thanks also to @audreystott for the amazing work.

I have tested the new installation within my personal space.

Note that SHPC is now stricter when it comes to the existance of registry paths: commands will fail if registry paths declared in the settings are non existent.

The trickiest is the user-specific one, line 81 of `install_shpc.sh`: `${USER_PERMANENT_FILES_PREFIX}/\$PAWSEY_PROJECT/\$USER/setonix/shpc_registry`. To avoid issues, I am now creating the path in line 68.

This implies that the script `spack_create_user_moduletree.sh` will have to be correctly run for users, otherwise they will get crashes if trying to use SHPC.

[EDIT]: also bringing in our 30x biocontainer recipes here for now, as the main SHPC automated registry is missing several aliases. To be ditched when the latter gets fixed.

Enjoy!

### CODE REVIEW

File: scripts/install_shpc.sh [hunk 1]
```diff
@@ -66,6 +64,9 @@ sed -i "s;/.*/python.*$;/bin/sh\n'''exec' & \"\$0\" \"\$@\"\n' ''';g" "${INSTALL
 export PATH="${INSTALL_PREFIX}/${shpc_install_dir}/bin":$PATH
 export PYTHONPATH="${INSTALL_PREFIX}/${shpc_install_dir}/lib/python${python_version_major}.${python_version_minor}/site-packages":$PYTHONPATH
 
+# need to create this registry directory, otherwise corresponding config command above will fail
+mkdir -p ${USER_PERMANENT_FILES_PREFIX}/\$PAWSEY_PROJECT/\$USER/setonix/shpc_registry
```
Comment (dipietrantonio) [inline scripts/install_shpc.sh:68]: Shouldn't it be `$` instead of `\$`? We want the value of the variable to be actually substituted I believe.

**Comment by dipietrantonio** [comment]
Hi Marco, as I asked in the "inline comment", shouldn't be the bash variables not escaped in that path, when creating the path?

File: scripts/install_shpc.sh (hunk 1 above)
Comment (marcodelapierre) [reply scripts/install_shpc.sh]: thanks, fixed!

**Comment by marcodelapierre** [comment]
thanks, fixed!

File: scripts/install_shpc.sh [hunk 2]
```diff
@@ -66,6 +64,9 @@ sed -i "s;/.*/python.*$;/bin/sh\n'''exec' & \"\$0\" \"\$@\"\n' ''';g" "${INSTALL
 export PATH="${INSTALL_PREFIX}/${shpc_install_dir}/bin":$PATH
 export PYTHONPATH="${INSTALL_PREFIX}/${shpc_install_dir}/lib/python${python_version_major}.${python_version_minor}/site-packages":$PYTHONPATH
 
+# need to create this registry directory, otherwise corresponding config command above will fail
+mkdir -p ${USER_PERMANENT_FILES_PREFIX}/$PAWSEY_PROJECT/$USER/setonix/shpc_registry
```
Comment (pelahi) [inline scripts/install_shpc.sh:68]: Don't we need to add the `DATE_TAG` here?

File: scripts/install_shpc.sh [hunk 3]
```diff
@@ -74,10 +75,10 @@ export PYTHONPATH="${INSTALL_PREFIX}/${shpc_install_dir}/lib/python${python_vers
...
 # singularity for containers
 shpc config set container_tech:singularity
-shpc config remove registry:\$root_dir/registry
-shpc config add "registry:${INSTALL_PREFIX}/${shpc_install_dir}/registry"
-shpc config add "registry:${INSTALL_PREFIX}/${shpc_install_dir}/pawsey_registry"
-shpc config add "registry:${USER_PERMANENT_FILES_PREFIX}/\$PAWSEY_PROJECT/\$USER/setonix/shpc_registry"
+shpc config remove registry https://github.com/singularityhub/shpc-registry
+shpc config add registry "${INSTALL_PREFIX}/${shpc_install_dir}/registry"
+shpc config add registry "${INSTALL_PREFIX}/${shpc_install_dir}/pawsey_registry"
+shpc config add registry "${USER_PERMANENT_FILES_PREFIX}/\$PAWSEY_PROJECT/\$USER/setonix/shpc_registry"
```
Comment (pelahi) [inline scripts/install_shpc.sh:81]: Again for user prefixes, we need the `DATE_TAG`.

**Review by pelahi — CHANGES_REQUESTED** [review]
Overall, seems fine but we need to add `DATE_TAG`s I think.

File: scripts/install_shpc.sh (hunk 2 above)
Comment (marcodelapierre) [reply scripts/install_shpc.sh:68]: thanks, done!

File: scripts/install_shpc.sh (hunk 3 above)
Comment (marcodelapierre) [reply scripts/install_shpc.sh]: thanks, done!

**Comment by marcodelapierre** [comment]
also merged latest `main` in, with all missing DATE_TAGs

**Comment by marcodelapierre** [comment]
[EDIT]: also bringing in our 30x biocontainer recipes here for now, as the main SHPC automated registry is missing several aliases. To be ditched when the latter gets fixed.

**Comment by dipietrantonio** [comment]
Hi Marco, thanks for the amazing contribution, very much appreciated. I have been testing the script `install_shpc.sh` but unfortunately some strange behaviour occurs. When the first shpc command is executed, line 75, a new shell is open and seems to wait for some input, but this is a wild guess, it needs to be investigated further. Here is what happens:

```
spack@nid001000:/scratch/pawsey0001/spack/marco/pawsey-spack-config> ./scripts/install_shpc.sh 
[...]
  Using cached MarkupSafe-2.1.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Collecting attrs>=17.4.0
  Using cached attrs-22.2.0-py3-none-any.whl (60 kB)
Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0
  Using cached pyrsistent-0.19.3-py3-none-any.whl (57 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2022.12.7-py3-none-any.whl (155 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.0.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (198 kB)
Collecting urllib3<1.27,>=1.21.1
  Using cached urllib3-1.26.14-py2.py3-none-any.whl (140 kB)
Collecting ruamel.yaml.clib>=0.2.6
  Using cached ruamel.yaml.clib-0.2.7-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (519 kB)
Installing collected packages: spython, charset-normalizer, urllib3, ruamel.yaml.clib, pyrsistent, MarkupSafe, idna, certifi, attrs, ruamel.yaml, requests, jsonschema, Jinja2, singularity-hpc
Successfully installed Jinja2-3.1.2 MarkupSafe-2.1.2 attrs-22.2.0 certifi-2022.12.7 charset-normalizer-3.0.1 idna-3.4 jsonschema-4.17.3 pyrsistent-0.19.3 requests-2.28.2 ruamel.yaml-0.17.21 ruamel.yaml.clib-0.2.7 singularity-hpc-0.0.57 spython-0.3.0 urllib3-1.26.14

[notice] A new release of pip available: 22.2.2 -> 23.0
[notice] To update, run: /scratch/pawsey0001/spack/marco/2023.02/software/linux-sles15-zen3/gcc-12.1.0/python-3.9.15-vs5k6iyeua3xb3oxj3nrawmazibajdgz/bin/python3.9 -m pip install --upgrade pip
SHPC commands start..
sh-4.4$ 

```

See how the prompt differs?

**Comment by marcodelapierre** [comment]
Hi Cristian,

where are you testing this?

I have a slightly modified version on Setonix software for the testing, which works:
```
INSTALL_GROUP=pawsey0001 bash /software/projects/pawsey0001/mdelapierre/testing/shpc-bioc-audrey/feb23/install-shpc-0.branch2.sh 
```

Just double-check the paths I have used...this should work:
```
mkdir -p /software/projects/$PAWSEY_PROJECT/$USER/testing/shpc-bioc-audrey/feb23
cd /software/projects/$PAWSEY_PROJECT/$USER/testing/shpc-bioc-audrey/feb23
cp -p /software/projects/pawsey0001/mdelapierre/testing/shpc-bioc-audrey/feb23/install-shpc-0.branch2.sh  .
cp -p /software/projects/pawsey0001/mdelapierre/testing/shpc-bioc-audrey/feb23/settings.sh .
INSTALL_GROUP=pawsey0001 bash ./install-shpc-0.branch2.sh 
```

Not sure yet at this stage about the differences between my run and your run.

**Comment by dipietrantonio** [comment]
@marcodelapierre and I had a videocall session and identified the issue in the sed commands, hence the above commit by Marco to fix the error. Unfortunately the issue persists:

```
spack@nid001002:/scratch/pawsey0001/spack/marco/pawsey-spack-config> head ../2023.02/pawsey/software/shpc/bin/shpc 
#!/bin/sh
'''exec' /bin/sh
'''exec' /scratch/pawsey0001/spack/marco/2023.02/software/linux-sles15-zen3/gcc-12.1.0/python-3.9.15-vs5k6iyeua3xb3oxj3nrawmazibajdgz/bin/python3.9 "$0" "$@" "$0" "$@"
' '''
' '''
# -*- coding: utf-8 -*-
import re
import sys
from shpc.client import run_shpc
if __name__ == '__main__':

```

**Comment by marcodelapierre** [comment]
Ok after testing with Cristian again, we're getting rid of the `sed` for the shebang; the pip installation process takes care of appropriate adjustments.

**Comment by dipietrantonio** [comment]
@marcodelapierre thanks for the fix! are you able to reply to the "request for change" of Pascal? you should be able to mark it as done.

**Review by dipietrantonio — APPROVED** [review]

**Comment by marcodelapierre** [comment]
@pelahi  I have addressed your requests, please review when you have the chance

**Review by pelahi — APPROVED** [review]

**Pull request merged**